pulsar/internal/connection.go

```diff
@@ -858,13 +858,14 @@ func (c *connection) Close() {
 		c.Lock()
 		cnx := c.cnx
 		c.Unlock()
-		c.changeState(connectionClosed)
+		c.changeState(connectionClosing)
```

Comment:
How does changing the state here to `connectionClosing` and then changing it to `connectionClosed` later fix the issue?

Reply:
This is not to say that the problem is solved, but first change the status to `connectionClosing` , the logic is more rigorous, I modified the description of the PR, it does not mean that the issue is fixed, but it is related to investigating this issue.

Reply:
If there is no intermediate `connectionClosing` state, other threads may start to create new connections in `GetConnection` before the current `cnx` is really closed. This is a mistake in itself. State transitions should be stricter. You can see that the network connection state transition of the java client is very rigorous.

If we can ensure that the state transition is more rigorous, set the state to `connectionClosing` before `cnx`  closed, which can prevent other threads to create a new connection. This may be able to solve the issue #676 that the old connection was not closed, and the new connection has initiated a request, even if this is not the root cause of the problem.